### DISCUSSION TITLE
With v2, how to have internal tools?

### DISCUSSION THREAD

**Discussion by lukpsaxo** [vercel/turborepo#8345] [Help]
### Summary

With v2, internal package dependencies in the root trigger breaking the cache.

We have alot of tools, most of which are used by developers, some of which are used during the build, but nothing used by turbo.

e.g. we have a stacktrace tool used by developers. its in a package in the monorepo - tools/stacktrace

right now we have a dependency on it in the root, then inside the stacktrace package.json we have:

```
    "bin": {
        "stacktrace": "./index.js",
```

and we can run the tool in the root with

```
yarn stacktrace
```

What is the best practice with turbo to avoid a tool breaking the cache and for it to be easy to use?

### Additional information

```js
So in order that a change to the stacktrace tool doesn't break all the turbo caches, the options seem to be:

1.
`npx stacktrace` - slower, doesn't use yarn so skips our env files
2.
`yarn -dlx @scope/stacktrace` - much slower
3.
Add a command to the root scripts:
`"stacktrace": "tools/stacktrace/node_modules/.bin/stacktrace",`
4.
Add a command to the stacktrace scripts package and then run `yarn workspace @scope/stacktrace stacktrace` to run the command (we could potentially do this and alias it in the root package.json)
```

**Comment by anthonyshew** [comment]
Hello and good question!

This change was made to more accurately reflect the state of your repository. If something is installed in the root of the repository and has changes, those changes invalidate the entire package graph. So, ultimately, this change was made for the sake of correctness.

If you do still feel strongly that you don't want the cache to miss when you make changes to the package, there are workarounds that don't involve adding it as a dependency to the Workspace root. The options you've listed all strike me as viable techniques, depending on the details of your repository.
